tokenserver-db-postgres/src/models.rs

```diff
@@ -363,6 +368,313 @@ impl TokenserverPgDb {
+            WHERE service = <service String>
+            AND email = <email String>
+            AND generation <=  <generation i64>,
+            AND COALESCE(keys_changed_at, 0) <= COALESCE(?, keys_changed_at, 0)
+            AND replaced_at IS NULL
+
+    */
+    async fn put_user(&mut self, params: params::PutUser) -> DbResult<results::PutUser> {
+        // As an added layer of safety, the `WHERE` clause ensures that concurrent updates
+        // don't accidentally move timestamp fields backwards in time. The handling of
+        // `keys_changed_at`can be problematic as we want to treat the default `NULL` as zero (0).
+        const QUERY: &str = r#"
+            UPDATE users
+                SET generation = $1,
+                    keys_changed_at = $2
+            WHERE service = $3
+                email = $4
+                generation <= $5
+                COALESCE(keys_changed_at, 0) <= COALESCE($6, keys_changed_at, 0)
+                replaced_at IS NULL
+        "#;
+
+        let mut metrics = self.metrics.clone();
+        metrics.start_timer("storage.put_user", None);
+
+        diesel::sql_query(QUERY)
+            .bind::<BigInt, _>(params.generation)
+            .bind::<Nullable<BigInt>, _>(params.keys_changed_at)
+            .bind::<Integer, _>(params.service_id)
+            .bind::<Text, _>(params.email)
+            .bind::<BigInt, _>(params.generation)
+            .bind::<Nullable<BigInt>, _>(params.keys_changed_at)
+            .execute(&mut self.conn)
+            .await
+            .map(|_| ())
+            .map_err(Into::into)
+    }
+
+    /**
+    Update the user record with the given uid and service id
+    marking it as 'replaced'. This is through updating the `replaced_at` field.
+
+        UPDATE users
+            SET replaced_at = <replaced_at i64>
+        WHERE service = <service i32>
+            AND uid = <uid i64>
+
+    */
+    async fn replace_user(
+        &mut self,
+        params: params::ReplaceUser,
+    ) -> DbResult<results::ReplaceUser> {
+        const QUERY: &str = r#"
+            UPDATE users
+                SET replaced_at = $1
+             WHERE service = $2
+                AND uid = $3
+        "#;
+
+        diesel::sql_query(QUERY)
+            .bind::<BigInt, _>(params.replaced_at)
+            .bind::<Integer, _>(params.service_id)
+            .bind::<BigInt, _>(params.uid)
+            .execute(&mut self.conn)
+            .await
+            .map(|_| ())
+            .map_err(Into::into)
+    }
+
+    /**
+    Update several user records with the given email and service id
+    marking them as 'replaced'. This is through updating the `replaced_at` field.
+    The `replaced_at` field should be null AND the `created_at` field should be earlier
+    than the `replaced_at`.
+
+        UPDATE users
+            SET replaced_at = <replaced_at i64>
+            WHERE service = <service i32>
+            AND uid = <uid i64>
+
```

Comment:
I don't think we need queries in docstrings, especially since we have raw queries in the code. This one's out of sync with the actual query in code -- I think that's always a risk when duplicating them.

The parameter type information can be useful but we get a form of that from diesel's `bind` calls.

Reply:
let's kill it anyway (sorry) -- I don't think the `TokenserverPgDb` specific methods would pop up in the IDE very often, if at all, as most usage of these methods is on the `Db` trait versions (and when they do you're just a hop to definition away to see the sql albeit w/ no formatting)

Apparently there's a [code plugin that syntax highlights sql](https://marketplace.visualstudio.com/items?itemName=qufiwefefwoyn.inline-sql-syntax) -- unfortunately it requires a `--sql` prefix to enable it per string, which is just an extra enough of annoyance to discourage its usage IMHO 🤷‍♂️